lib/ecto/schema.ex

```diff
@@ -1771,11 +1771,35 @@ defmodule Ecto.Schema do
 
   @doc false
   def __field__(mod, name, type, opts) do
+    type = maybe_to_parameterized(type, name, mod)
```

Comment:
Instead of doing an early pass, change `check_field_type!` to return the type and we do all traversal there. Especially if we are going with `field :foo, Ecto.Enum, values: [...]`, since `check_field_type!` already has logic to validate custom types.